src/packtPublishingFreeEbook.py

```diff
@@ -145,15 +145,19 @@ def grab_ebook(self, api_client):
         [user_data] = user_response.json().get('data')
         user_id = user_data.get('id')
 
+        product_response = api_client.get(PACKT_PRODUCT_SUMMARY_URL.format(product_id=product_id))
+        self.book_data = {'id': product_id, 'title': product_response.json()['title']}\
+            if product_response.status_code == 200 else None
+
+        if any(product_id == book['id'] for book in self.get_single_page_books_data(api_client, 0)):
```

Comment:
Why not `if any(product_id == book['id'] for book in self.get_all_books_data(api_client)):`? Currently you're checking only for the first pagination page, with proposed change we'd check all user's products.

Reply:
because I want to reduce the api call and to boost speed. but right that is more stable way.

Reply:
Solving ReCaptcha is what slows us down - we could parallelize solving ReCaptcha and other operation, so fetching all user's books data won't slow us down.

Reply:
but if you have really many book. it would make the slow down too.
also IMO, we can't parallelize the solve captcha & another process part.
only can check the claim before the solving. it means this action could be `serial`

maybe I'm getting wrong. I just focused this `grab_ebook` part.
if we can run parallelize another part like download all ebooks and claim new books.
you might be right, but imo, we need cache in this case.

Reply:
Currently we make 3 queries to Packt API before claiming query (and we have to have ReCapcha solution to make this query). Even if somebody has - let's say - 1000 products in the account (and I doubt most users have that many) than we would have to make 3 + (1 + 1000 / 25) = 44 queries (assuming pagination page size of 25). Solving ReCaptcha sometimes takes even more than 2 minutes (timeouts happen), so it means that average API query would have to take 120 / 44 = 2.(72) seconds to slow down the paralleled script (which is unlikely, I'd assume that average query takes a fraction of second).

Reply:
hm. okay, I got why we talk about another part.
basically, I want to prevent to reclaim the free book, that original process always try to solve the captcha when entered claim page.
so I thought, this process should be serialize, first fetch the ebook list that I already have then check that I have the ebook or not, then try to solve the captcha to claim the free book.
why I thought this, because solving the captcha isn't free.

but your opinion is quite different, if I got, you said just fetch the ebook list, and try to solve the captcha at the same time (because these sections are a bottle neck), then try to claim the free book.